Userland/Libraries/LibGfx/AntiAliasingPainter.cpp

```diff
@@ -18,148 +18,135 @@
+        start_point.set_y(round_to<int>(start_point.y()));
+        return fill_rect(Gfx::FloatRect(start_point, { length, float(int_thickness) }), color);
+    }
+    if (mapped_from.x() == mapped_to.x()) {
+        auto start_point = (mapped_from.y() < mapped_to.y() ? mapped_from : mapped_to).translated(-int_thickness / 2, 0);
+        return fill_rect(Gfx::FloatRect(start_point, { float(int_thickness), length }), color);
     }
 
-    rotated_rectangle_reference_coords = build_rotated_rectangle(draw_direction, corrected_thickness);
-
-    auto plot = [&](int x, int y, float c) {
-        // ignore rotation if rectangle is fairly small to reduce overhead
-        if (is_straight_line || corrected_thickness < 4) {
-            m_underlying_painter.fill_rect(IntRect::centered_on({ x, y }, size), color.with_alpha(color.alpha() * c));
-        } else if (min(AK::abs(mapped_from.distance_from({ x, y })), AK::abs(mapped_to.distance_from({ x, y }))) >= drawing_edge_offset || AK::abs(mapped_from.distance_from(mapped_to)) < drawing_edge_offset) {
-            // don't draw if we are close to the edge but draw if the whole line is shorter than allowed edge distance
-            m_rotated_rectangle_path.clear();
-            m_rotated_rectangle_path.move_to({ x + rotated_rectangle_reference_coords.p1().x(), y + rotated_rectangle_reference_coords.p1().y() });
-            m_rotated_rectangle_path.line_to({ x + rotated_rectangle_reference_coords.p2().x(), y + rotated_rectangle_reference_coords.p2().y() });
-            m_rotated_rectangle_path.line_to({ x + rotated_rectangle_reference_coords.p3().x(), y + rotated_rectangle_reference_coords.p3().y() });
-            m_rotated_rectangle_path.line_to({ x + rotated_rectangle_reference_coords.p4().x(), y + rotated_rectangle_reference_coords.p4().y() });
-
-            m_rotated_rectangle_path.close();
-            m_underlying_painter.fill_path(m_rotated_rectangle_path, color.with_alpha(color.alpha() * c));
-        }
+    // The painting only works for the positive XY quadrant (because that is easier).
+    // So flip things around until we're there:
+    bool flip_x = false;
+    bool flip_y = false;
+    if (mapped_to.x() < mapped_from.x() && mapped_to.y() < mapped_from.y())
+        swap(mapped_to, mapped_from);
+    if ((flip_x = mapped_to.x() < mapped_from.x()))
+        mapped_to.set_x(2 * mapped_from.x() - mapped_to.x());
+    if ((flip_y = mapped_to.y() < mapped_from.y()))
+        mapped_to.set_y(2 * mapped_from.y() - mapped_to.y());
+
+    // FIXME: SVG tactical counter hack
+    // When painting an any curve the SVG painting asks for loads of 0.4px long lines,
+    // if you actually draw 0.4px longs lines it looks broken.
+    // This is because the SVG code relies on a bug from the old line rendering code,
+    // Where for any non-axis aligned line, the minimum length was around 5px long.
+    // This hacks around that...
+    if (length > 0.1f && length < 1.0f)
+        length = 3;
```

Comment:
If this is a SVG hack, shouldn't this go into the SVG code then? Or at least have an alternative line drawing implementation for SVG so we do not bother our generic (as correct as possible) implementation with this?

Reply:
This does not break anything else that uses these lines right now (which I think is just PixelPaint), because they don't/can't ask for a line < 1px long.

I did have a look into fixing the SVG code, but it's a little less familiar to me, so I didn't see something obvious (but I'll have another look). 

Fundamentally though the way we draw curves for SVG is kinda hack in itself, you can't really draw a curve as loads of subpixel AA lines and get a nice antialiased result.

Reply:
Didn't find a fix, but I've found a new hack that gives shockingly good results